cmd/juju/application/utils/utils.go

```diff
@@ -118,38 +153,87 @@ func filterResources(
 }
 
 // shouldUpgradeResource reports whether we should upload the metadata for the given
-// resource.  This is always true for resources we're adding with the --resource
-// flag. For resources we're not adding with --resource, we only upload metadata
-// for charmstore resources.  Previously uploaded resources stay pinned to the
-// data the user uploaded.
-func shouldUpgradeResource(newCharmURL *charm.URL, res charmresource.Meta, uploads map[string]string, current map[string]resources.Resource) (bool, error) {
-	// Always upload metadata for resources the user is uploading during
-	// upgrade-charm.
-	if _, ok := uploads[res.Name]; ok {
-		logger.Tracef("%q provided to upgrade existing resource", res.Name)
+// resource.
+//
+// Upgrade when:
+//  1. Resources specified as a local file on the cli to be uploaded.
+//  2. Resources specified as resource revision on the cli, which are
+//     not currently in use.
+//  3. Upstream has a new resource available than what is currently used.
+//  4. A new upstream resource is available.
+//
+// Caveat: Previously uploaded resources stay pinned to the data the user uploaded.
+func shouldUpgradeResource(
+	resName string,
+	providedResources map[string]string,
+	current map[string]resources.Resource,
+	available map[string]charmresource.Resource,
+) (bool, error) {
+
+	cur, curFound := current[resName]
+	providedResource, providedResourceFound := providedResources[resName]
+
+	if providedResourceFound {
+		providedResourceRev, err := strconv.Atoi(providedResource)
+		if err == nil && curFound && cur.Revision == providedResourceRev && cur.Origin == charmresource.OriginStore {
+			// A revision refers to resources in a repository. If the specified revision
+			// is already uploaded, nothing to do.
+			logger.Tracef("provided revision of %s resource already loaded", resName)
+			return false, nil
+		}
+		logger.Tracef("%q provided to upgrade existing resource", resName)
 		return true, nil
 	}
 
-	cur, ok := current[res.Name]
-	if !ok {
+	if !curFound {
 		// If there's no information on the server, there might be a new resource added to the charm.
-		if newCharmURL.Schema == charm.Local.String() {
-			return false, errors.NewNotValid(nil, fmt.Sprintf("new resource %q was missing, please provide it via --resource", res.Name))
-		}
-
-		logger.Tracef("resource %q does not exist in controller, so it will be uploaded", res.Name)
+		logger.Tracef("resource %q does not exist in controller, so it will be uploaded", resName)
```

Comment:
s/uploaded/upgraded?

Reply:
The charm has introduced a new resource that the application hasn't used before.  The user hasn't listed a file on the cli to use for it which would be uploaded. Maybe downloaded or keep the same?

Reply:
I think just tweak the message to avoid any ambiguity